jbpm-flow/src/main/java/org/jbpm/workflow/instance/node/RuleSetNodeInstance.java

```diff
@@ -65,83 +75,112 @@
+                String decision = resolveVariable(ruleSetNode.getDecision());
+
+                DMNRuntime runtime = ((KieSession) kruntime).getKieRuntime(DMNRuntime.class);
+                DMNModel dmnModel = runtime.getModel(namespace, model);
+                if (dmnModel == null) {
+                    throw new IllegalArgumentException("DMN model '" + model + "' not found with namespace '" + namespace + "'");
+                }
+                DMNResult dmnResult = null;
+                DMNContext context = runtime.newContext();
+
+                for (Entry<String, Object> entry : inputs.entrySet()) {
+                    context.set(entry.getKey(), entry.getValue());
+                }
+
+                if (decision != null && !decision.isEmpty()) {
+                    dmnResult = runtime.evaluateDecisionByName(dmnModel, decision, context);
+                } else {
+                    dmnResult = runtime.evaluateAll(dmnModel, context);
+                }
+
+                if (dmnResult.hasErrors()) {
+                    String errors = dmnResult.getMessages(Severity.ERROR).stream()
+                            .map(message -> message.toString())
+                            .collect(Collectors.joining(", "));
+
+                    throw new RuntimeException("DMN result errors:: " + errors);
+                }
+                processOutputs(dmnResult.getContext().getAll());
                 triggerCompleted();
-        	}
-    	}
+            } else {
+
+                // first set rule flow group
+                setRuleFlowGroup(resolveRuleFlowGroup(ruleSetNode.getRuleFlowGroup()));
+
+                //proceed
+                for (Entry<String, Object> entry : inputs.entrySet()) {
+                    String inputKey = getRuleFlowGroup() + "_" + getProcessInstance().getId() + "_" + entry.getKey();
+
+                    factHandles.put(inputKey, kruntime.insert(entry.getValue()));
+                }
+
+                if (actAsWaitState()) {
+                    addRuleSetListener();
+                    ((InternalAgenda) getProcessInstance().getKnowledgeRuntime().getAgenda())
+                            .activateRuleFlowGroup(getRuleFlowGroup(), getProcessInstance().getId(), getUniqueId());
+
+                } else {
+                    ((InternalAgenda) getProcessInstance().getKnowledgeRuntime().getAgenda())
+                            .activateRuleFlowGroup(getRuleFlowGroup(), getProcessInstance().getId(), getUniqueId());
+
+                    ((KieSession) getProcessInstance().getKnowledgeRuntime()).fireAllRules();
+
+                    removeEventListeners();
+                    retractFacts();
+                    triggerCompleted();
+                }
+            }
+        } catch (Exception e) {
+            if(!handleException(e, false)) {
+                throw new WorkflowRuntimeException(this, getProcessInstance(), "Unable to execute Action: " + e.getMessage(), e);
+            }
```

Comment:
can we move this logic to handlerException and in case it's not handled simply throw exception?